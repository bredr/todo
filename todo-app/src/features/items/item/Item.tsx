import { grpc } from "@improbable-eng/grpc-web";
import { Fab } from "@material-ui/core";
import Button from "@material-ui/core/Button";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import { makeStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import DeleteIcon from "@material-ui/icons/Delete";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { ID, Item as ItemType } from "../../../proto/todo_pb";
import { Todo } from "../../../proto/todo_pb_service";
import { getItemsAction } from "../../pagination/Pagination";
import toaster from "../../toaster/state/actions";
import { editItem, removeItem } from "../state/actions";

const useStyles = makeStyles(theme => ({
  fab: {
    margin: theme.spacing(1)
  },
  card: {
    minWidth: 275
  },
  bullet: {
    display: "inline-block",
    margin: "0 2px",
    transform: "scale(0.8)"
  },
  title: {
    fontSize: 14
  },
  pos: {
    marginBottom: 12
  }
}));

const deleteItemAction = (id: string) => {
  const request = new ID();
  request.setId(id);
  return removeItem({
    methodDescriptor: Todo.RemoveItem,
    request,
    onMessage: () => getItemsAction(),
    onEnd: (code, message) =>
      code === grpc.Code.OK
        ? toaster.addSuccess("Deleted")
        : toaster.addError(message)
  });
};

export const Item: React.FC<ItemType.AsObject> = props => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const { description, daystocomplete, due, id } = props;

  function onEdit() {
    dispatch(editItem(id));
  }

  function onDelete() {
    dispatch(deleteItemAction(id));
  }

  return (
    <Card className={classes.card}>
      <CardContent>
        <Typography
          className={classes.title}
          color="textSecondary"
          gutterBottom={true}
        >
          {description}
        </Typography>
        <Typography variant="body2" component="p">
          Days to complete: {daystocomplete}
          <br />
          {due
            ? "Due:" + new Date(due.seconds * 1000).toLocaleDateString()
            : ""}
        </Typography>
      </CardContent>
      <CardActions>
        <Button size="small" onClick={onEdit} className={classes.fab}>
          Edit
        </Button>
        <Fab
          size="small"
          aria-label="delete"
          onClick={onDelete}
          className={classes.fab}
        >
          <DeleteIcon />
        </Fab>
      </CardActions>
    </Card>
  );
};
