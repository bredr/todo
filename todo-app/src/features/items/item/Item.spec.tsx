import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { Item } from "./Item";

const mockStore = configureStore([]);

describe("Item", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const props = {
      description: "d",
      daystocomplete: 1,
      due: { seconds: 1000000, nanos: 0 },
      id: "1"
    };
    const store = mockStore();
    const { container } = render(
      <Provider store={store}>
        <Item {...props} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("dispatches delete", () => {
    const props = {
      description: "d",
      daystocomplete: 1,
      due: { seconds: 1000000, nanos: 0 },
      id: "1"
    };
    const store = mockStore();
    const { getByLabelText } = render(
      <Provider store={store}>
        <Item {...props} />
      </Provider>
    );
    fireEvent.click(getByLabelText("delete"));
    expect(store.getActions()[0]).toEqual(
      expect.objectContaining({ type: "GRPC_WEB_REQUEST" })
    );
  });

  it("dispatches edit", () => {
    const props = {
      description: "d",
      daystocomplete: 1,
      due: { seconds: 1000000, nanos: 0 },
      id: "1"
    };
    const store = mockStore();
    const { getByLabelText } = render(
      <Provider store={store}>
        <Item {...props} />
      </Provider>
    );
    fireEvent.click(getByLabelText("edit"));
    expect(store.getActions()[0]).toEqual(
      expect.objectContaining({
        error: undefined,
        meta: undefined,
        payload: "1",
        type: "EDIT_ITEM"
      })
    );
  });
});
