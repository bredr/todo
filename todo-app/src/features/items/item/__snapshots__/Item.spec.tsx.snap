// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-2 MuiPaper-rounded"
>
  <div
    class="MuiCardContent-root"
  >
    <p
      class="MuiTypography-root makeStyles-title-4 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
    >
      d
    </p>
    <p
      class="MuiTypography-root MuiTypography-body2"
    >
      Days to complete: 
      1
      <br />
      Due:1/12/1970
    </p>
  </div>
  <div
    class="MuiCardActions-root MuiCardActions-spacing"
  >
    <button
      aria-label="edit"
      class="MuiButtonBase-root MuiButton-root makeStyles-fab-1 MuiButton-text MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Edit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      aria-label="delete"
      class="MuiButtonBase-root MuiFab-root makeStyles-fab-1 MuiFab-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
